[uss_qualifier] Adding scenario for SCD0090.

sahlman-nasa opened the pull request:
This is the NASA submitted scenario to test ASTM requirement SCD0090.

seungmanlee-nasa commented:
@barroco Thanks for your comments. I have a question about the benefit 1. Each requirement in the test suite table has multiple scenarios.: https://github.com/interuss/monitoring/blob/main/monitoring/uss_qualifier/suites/astm/utm/f3548_21.md. I was wondering if we execute all of the multiple scenarios for each requirement during testing (in this case test scenarios will be executed multiple times) or I was wondering if each test scenario is still executed only once even though the test scenario is listed multiple times for other requirements.

BenjaminPelletier commented:
> Each requirement in the test suite table has multiple scenarios.

Many requirements are tested in only a single scenario each while other requirements are tested in multiple scenarios.

> I was wondering if all of the multiple test scenarios for each requirement are executed sequentially during testing

Test scenarios are executed in the order defined by their container (e.g., [test suite](https://github.com/interuss/monitoring/tree/main/monitoring/uss_qualifier#architecture)).  Execution does not proceed according to requirements.

> (in this case, test scenarios will be executed multiple times for other requirements as well)

No execution proceeds according to requirements.  Test scenarios are sometimes executed multiple time by action generators in response to the content of resources provided to the test configuration.  For example, if three USSs are provided in a resource and a test scenario has two roles, then we might want to execute that test scenario 6 times so that every combination of 2 USSs can be used to fill those roles.

> or I was wondering if each test scenario is still executed only once even though the test scenario is listed multiple times for other requirements.

Each test scenario is executed according to [architecture](https://github.com/interuss/monitoring/tree/main/monitoring/uss_qualifier#architecture) in the test configuration, none of which is determined by requirements.  The summarization of listing test scenarios next to requirements is essentially an index of test scenarios according to what requirements they are capable of checking (test suite documentation [is generated automatically](https://github.com/interuss/monitoring/blob/main/monitoring/uss_qualifier/suites/README.md#documentation))

sahlman-nasa commented:
@barroco 
>     1. As mentioned on [Slack](https://interuss.slack.com/archives/C05N1MWCVPA/p1710192414246739?thread_ts=1709925610.135029&cid=C05N1MWCVPA), we will want to extend an existing scenario such as [nominal_planning/conflict_equal_priority_not_permitted](https://github.com/interuss/monitoring/blob/main/monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/conflict_equal_priority_not_permitted.md). My expectation is that it would cover those test cases and we should be able to augment it with new checks. Have you had a change to look at it?
>        Two benefits would be to:
>        
>        1. Verify multiple requirements at the same time and not create a test scenario which would repeat existing steps and will impact unnecessarily the overall test run duration.
>        2. Wait durations could be removed by checking notifications at the end of the test scenario allowing to batch the checks of the multiple notifications later in the run.

Yes, I have looked at that test case, and I do see how this test for SCD0090 could be folded into it. We could put a check for conflict notifications at the end of the test, but how do I specify which notifications to look for? The conflict_equal_priority_not_permitted test includes several situations which should result in a number of conflict notifications via SCD0090 and SCD0095. If I wanted to check for notifications from the test section "Attempt to plan flight into conflict test case" would I just say to look for a notification to Flight 1 from this section? Would I also need to add a step at the end of the "Attempt to plan flight into conflict test case" to record the time that the verification check on the flight 1 OI was performed so that it could be compared to the timestamp on the corresponding conflict notification?

> 
>     2. Regarding the diagram, I think this is great documentation of the overall approach and capture well the general approach. Therefore, may I kindly ask you to create a PR for it into this [repository](https://github.com/interuss/tsc) in the [investigations folder alongside the design document](https://github.com/interuss/tsc/tree/main/investigations/user_notification) ? Please provide the plantuml source along the rendered png to allow further edition.

I did add this diagram to the repo you specified. I waited a little while because we were discussing some further enhancments to our diagrams.

sahlman-nasa commented:
Ok I have modified it to use the existing test instead. I managed to do something to get my git state messed up though I did eventually get it synched properly again (I tried to pull and merge with main - maybe I shouldn't have done that). I don't know if I may have introduced any weird statuses into this. 

I added a test fragment to record the time and flight status. Let me know if this is ok or needs to be modified.
I also changed the validate conflict notification test fragment so that it could be used for SCD0090 or SCD0095 because they are very similar with a minor difference. Please let me know if this is ok, or if I should split this into two test fragments.

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/conflict_equal_priority_not_permitted.md:
```diff
@@ -131,6 +131,9 @@ per **[astm.f3548.v21.SCD0035](../../../../../requirements/astm/f3548/v21.md)**.
 
+### Record time and flight status test step
+#### [Record time and flight status](test_steps/record_status_for_notification_check.md)
+Record the time and flight status data for use in verifying notifications later.
```
Ok, but are you saying that this should be removed here? If so do you mean all lines 134 - 136? Also if so, do we need to note that this is the the moment to mark the time in some other way, or how do we ensure that the time and flight status will be properly recorded for later reference?

sahlman-nasa replied:
Removed.

sahlman-nasa commented on monitoring/uss_qualifier/requirements/interuss/tested_uss/hosted_instance.md:
```diff
@@ -0,0 +1,11 @@
+
+## Requirements
+
+### <tt>ExposeInterface</tt>
```
Done.

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/conflict_equal_priority_not_permitted.md:
```diff
@@ -167,6 +170,10 @@ As such the change to Flight 1c should be rejected per **[astm.f3548.v21.SCD0040
 
+### Record time and flight status test step
+#### [Record time and flight status](test_steps/record_status_for_notification_check.md)
+Record the time and flight status data for use in verifying notifications later.
```
Done

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/test_steps/record_status_for_notification_check.md:
```diff
@@ -0,0 +1,3 @@
+# Record status for notification check test step fragment
```
Removed

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/test_steps/validate_conflict_notification_to_user.md:
```diff
@@ -0,0 +1,60 @@
+## 🛑 Tested USS conflict notifications logs retrievable check
+
+If the query to tested USS fails or uss_qualifier is otherwise unable to retrieve the conflict notifications, the tested USS provider does not meet
+**[interuss.tested_uss.hosted_instance.ExposeInterface](../../../../../../requirements/interuss/tested_uss/hosted_instance.md)**.
```
Done

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/test_steps/validate_conflict_notification_to_user.md:
```diff
@@ -0,0 +1,60 @@
+
+## ⚠️ Expect conflict notification sent check
+As per **[astm.f3548.v21.SCD0090](../../../../../../requirements/astm/f3548/v21.md)**, a conflict notification should be
+sent by the USS to the user of a new or modified operational intent, and per **[astm.f3548.v21.SCD0095](../../../../../../requirements/astm/f3548/v21.md)**, a conflict notification should be
```
Added more checks for these cases.

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/test_steps/validate_conflict_notification_to_user.md:
```diff
@@ -0,0 +1,60 @@
+a high confidence that the test driver correctly verifies if a notification was sent by tested_uss.
+
+## 🛑 Conflict notification data is valid check
+If data is not correct or exceeds the 99 percentile wait time of 7 seconds, the tested USS provider does not meet **[astm.f3548.v21.SCD0090](../../../../../../requirements/astm/f3548/v21.md)** or **[astm.f3548.v21.SCD0095](../../../../../../requirements/astm/f3548/v21.md)**.
```
Removed references to SCD0095.

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/test_steps/validate_user_conflict_notification_with_another_flight.md:
```diff
@@ -0,0 +1,56 @@
+
+This step verifies that when creating or modifying an operational intent, the tested USS sent the required notification to the user of the new or modified conflicting operational intent.
+
+## 🛑 Conflict notifications logs retrievable check
```
OK, though I'm not entirely clear on what constitutes a stop vs. a warning, unless it has to do with a condition that indicates that something in the scenario environment is not working as expected vs a test condition that failed because it didn't meet its specification (but the environment is still working as expected).

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/test_steps/validate_user_conflict_notification_with_another_flight.md:
```diff
@@ -0,0 +1,56 @@
+If exactly one notification to the user per detected conflict is not present, this check will fail per **[astm.f3548.v21.SCD0090](../../../../../../requirements/astm/f3548/v21.md)**
+
+## 🛑 Performance evaluation of conflict notification data check
+If notification exceeds the 99 percentile wait time of 7 seconds, the tested USS provider does not meet **[astm.f3548.v21.SCD0090](../../../../../../requirements/astm/f3548/v21.md)**.
```
Ok, I moved up some of the text from the Notes section.

sahlman-nasa commented on monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/test_steps/validate_user_conflict_notification_with_another_flight.md:
```diff
@@ -0,0 +1,56 @@
+## 🛑 Performance evaluation of conflict notification data check
+If notification exceeds the 99 percentile wait time of 7 seconds, the tested USS provider does not meet **[astm.f3548.v21.SCD0090](../../../../../../requirements/astm/f3548/v21.md)**.
+
+#### Note
```
I patterned this after a similar test step fragment I found that also included a Notes section like this. I thought it would be useful to include similar details here, but if not, I'll go ahead and remove it. I'm not sure if there's a way to make this section like a comment block, but if so, maybe I could add it back as that.

sahlman-nasa commented:
Punam brought up an interesting point today. She pointed out that the way we're testing this with respect to SCD0090, the new and modified flight (Operational Intent) is rejected and never actually created. The wording of SCD0090 is that the user of a "created" or "modified" flight must be notified of the conflict. Because the cases that we are looking at here are rejected, they are not truly "created" or "modified". The question is does that matter? Is it good enough to "attempt to" create or modify a flight and does that satisfy SCD0090? She was further concerned that a rejection itself would likely include a reason for rejection and thus that a notification (via SCD0090) would not be required or sent. What do you think about this? If this doesn't really test SCD0090, then we may have to redo this scenario (including the diagram) in a manner that does test it.

sahlman-nasa commented:
> 
>  An alternative would be to create a new scenario covering the case `conflict with equal priority, permitted` 

Early on when we started talking about supporting the Key Site demo and realized that we would have to create scenarios for equal-priority, no-conflict cases we were wondering how relevant requirements SCD0090 and SCD0095 would be. Still it seems they can be tested using Nonconforming flights as was done in this scenario so I can create a new scenario involving a Nonconforming case to force a conflict.

I have a couple of questions though. I notice that in the step Successful transition to non-conforming state check it says "If the USS indicates a conflict, this check will fail." Isn't that wrong? The point of the step is to create a conflict with a Nonconforming flight. It seems like it should say the opposite.

Is there a preference grouping scenarios in the directory structure? I notice that in nominal_planning there are subfolders for the individual scenarios, but, for example, in off_nominal_planning there are multiple scenarios within that folder and no scenarios in subfolders.

seungmanlee-nasa commented:
> Indeed, however I don't think we need a new scenario, this [step](https://github.com/interuss/monitoring/blob/4e6f647308e3edeea7a7da8b1bede6472c1cad4c/monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/conflict_equal_priority_not_permitted.md#attempt-to-modify-activated-flight-1-in-conflict-with-nonconforming-flight-2-test-step) should trigger a notification if the change has been accepted ? Please, note that both following USS behaviours are acceptable:
> a. the USS decides to proceed with the change and the conflict with the non-conforming operational intent is performed.
> b. the USS decides to reject the change and not proceed.
> Notification is required to be sent to the user in the case (a) only.

Hi Michale, Even though the USS decides to reject the change and not proceed for some reasons, shouldn't the USS send the notification of the conflict with non-conforming OI to the user based on SCD0090?

seungmanlee-nasa commented:
Another comment on this scenario is that it would be better to make test step fragment MD files for the test checks listed at the same level of the test step, such as "Successful modification or rejection check" and "Failure check," as other test scenarios for consistency. 
Also, it would be better to rename the "Validate tested USS conflict notifications to user", "Validate Flight 1 sharing", and "Validate tested USS conflict notification to user for attempted new flight" to indicate whether they are a test step, test case, or test check.

sahlman-nasa commented:
> > Early on when we started talking about supporting the Key Site demo and realized that we would have to create scenarios for equal-priority, no-conflict cases we were wondering how relevant requirements SCD0090 and SCD0095 would be. Still it seems they can be tested using Nonconforming flights as was done in this scenario so I can create a new scenario involving a Nonconforming case to force a conflict.
> 
> Indeed, however I don't think we need a new scenario, this [step](https://github.com/interuss/monitoring/blob/4e6f647308e3edeea7a7da8b1bede6472c1cad4c/monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/conflict_equal_priority_not_permitted.md#attempt-to-modify-activated-flight-1-in-conflict-with-nonconforming-flight-2-test-step) should trigger a notification if the change has been accepted ? Please, note that both following USS behaviours are acceptable: a. the USS decides to proceed with the change and the conflict with the non-conforming operational intent is performed. b. the USS decides to reject the change and not proceed. Notification is required to be sent to the user in the case (a) only.

I had considered this case, but I'm not sure it adequately tests SCD0090. Since this requirement is testing a new or modified flight in conflict, we would have to consider checking flight 1 since that is the one managed by the USS under test. It does not test the "new flight" part of the requirement as no flight is created into conflict (only modified). It also relies on a flight, already in conflict, to be allowed to be modified to a new state that is still in conflict. In this equal priority environment it would seem to give an additional reason that the USS could reject this modification and render the test inconclusive. If instead we take two flights that are not in conflict and modify one of them to be nonconforming and in conflict with the other, I think that would rely only on the USS to support the CMSA (and SCD) role, and I think more likely to produce an outcome. This is why I think a new scenario may be justified for this case.

seungmanlee-nasa commented:
> > Indeed, however I don't think we need a new scenario, this [step](https://github.com/interuss/monitoring/blob/4e6f647308e3edeea7a7da8b1bede6472c1cad4c/monitoring/uss_qualifier/scenarios/astm/utm/nominal_planning/conflict_equal_priority_not_permitted/conflict_equal_priority_not_permitted.md#attempt-to-modify-activated-flight-1-in-conflict-with-nonconforming-flight-2-test-step) should trigger a notification if the change has been accepted ? Please, note that both following USS behaviours are acceptable:
> > a. the USS decides to proceed with the change and the conflict with the non-conforming operational intent is performed.
> > b. the USS decides to reject the change and not proceed.
> > Notification is required to be sent to the user in the case (a) only.
> 
> Hi Michael, Even though the USS decides to reject the change and not proceed for some reason, shouldn't the USS send the notification of the conflict with non-conforming OI to the user based on SCD0090?

As described in the description of the test step "This modification results in a conflict between the two equal priority flight that already existed before the modification was initiated. **While this modification is expected to be accepted by the tested USS in general, the rejection of the modification does not constitute a violation of a requirement.** However, the modification request must not result in a failure per [interuss.automated_testing.flight_planning.ExpectedBehavior].", the USS may reject the modification for some reason. However, SCD0090 does NOT say that USS is not required to send a notification of the conflict with another OI. So, I believe that the managing USS should send the conflict notification even though the USS rejects the new or modified OI due to a conflict with a non-conforming OI based on the SCD0090 (although the USS can accept the modification request in general as described). Any thoughts?

sahlman-nasa commented:
> Another comment on this scenario is that it would be better to make test step fragment MD files for the test checks listed at the same level of the test step, such as "Successful modification or rejection check" and "Failure check," as other test scenarios for consistency. Also, it would be better to rename the "Validate tested USS conflict notifications to user", "Validate Flight 1 sharing", and "Validate tested USS conflict notification to user for attempted new flight" to indicate whether they are a test step, test case, or test check.

I see what you're saying about the check steps. It does seem that all the other checks are in the test step fragment files. However, this was preexisting before I made my modifications so I think this would be up to the InterUSS side to make these changes. With regard to your second comment, there does seem to be a standard that not every heading is listed as a "test case", "test step", or "test check". It seems that a "test case" is composed of one or more "test steps" and each test step contains one or more "substeps" which don't include a specific identifier. Steps such as "Validate Flight 1 sharing" would be this type of "substep" and are also preexisting; however I would agree that at least "Validate tested USS conflict notifications to user" should be labeled as a "test step". I think that "Validate tested USS conflict notification to user for attempted new flight" is consistent with other similar headings. If we want to assign a label to this level of headings, I would think that InterUSS would need to establish a standard for this. This is based on my admittedly limited understanding thus far though.

sahlman-nasa commented:
I have modified this scenario to use a nonconforming flight to create conflicts and separated the scenario back into a new file since I didn't feel it was compatible with the same priority no conflict scenario. Even if this isn't a useful for the upcoming Key Site event we would still like to get some feedback/suggestions on this approach for testing SCD0090. Thanks!

sahlman-nasa commented:
I think we are no longer pursuing this at this time so it can be closed for now.
